### PR TITLE
dai-zephyr: check max block count supported by DMA

### PR DESCRIPTION
**Pull request by abonislawski**
check max block count supported by DMA to prepare a correct config for DMA

Fixes: https://github.com/thesofproject/sof/issues/6709

OPEN: currently it is recalculating block count to max supported but it could also return error and recalculate to 1 only for HDA

### CODE REVIEW

File: src/audio/dai-zephyr.c [hunk 1]
```diff
@@ -452,7 +452,7 @@ static int dai_playback_params(struct comp_dev *dev, uint32_t period_bytes,
 		*local_buf = buffer_acquire(dd->local_buffer);
```
Comment (kv2019i) [inline src/audio/dai-zephyr.c:452]: The commit message could be a bit more descriptive why this change is done.

File: src/audio/dai-zephyr.c [hunk 2]
```diff
@@ -487,6 +487,20 @@ static int dai_playback_params(struct comp_dev *dev, uint32_t period_bytes,
...
+		if (err < 0) {
+			comp_err(dev, "dai_playback_params(): could not get dma attr max block count, err = %d", err);
+			goto out;
+		}
+
+		if (max_block_count < period_count) {
+			comp_info(dev, "dai_playback_params(): block count = %d not supported by DMA", period_count);
+			buf_size = period_count * period_bytes;
+			period_count = max_block_count;
+			period_bytes = buf_size / period_count;
```
Comment (kv2019i) [inline src/audio/dai-zephyr.c:501]: Maybe add an assert for the case where this does not divide evenly? Should never happen of course.

File: src/audio/dai-zephyr.c [hunk 3]
```diff
@@ -616,6 +630,20 @@ static int dai_capture_params(struct comp_dev *dev, uint32_t period_bytes,
...
+		if (err < 0) {
+			comp_err(dev, "dai_capture_params(): could not get dma attr max block count, err = %d", err);
+			goto out;
+		}
+
+		if (max_block_count < period_count) {
+			comp_info(dev, "dai_capture_params(): block count = %d not supported by DMA", period_count);
+			buf_size = period_count * period_bytes;
+			period_count = max_block_count;
+			period_bytes = buf_size / period_count;
```
Comment (kv2019i) [inline src/audio/dai-zephyr.c:644]: Ditto for the assert.

**Review by kv2019i — COMMENTED** [review]
Mostly good, a few minor comments inline. I think this is at least better than having HDA specific code in dai-zephyr.c.

File: src/audio/dai-zephyr.c (hunk 2 above)
Comment (lyakh) [reply src/audio/dai-zephyr.c:501]: hmm, maybe there's some implicit rule that I'm not aware of, but by just looking at this, it might well not divide evenly. We enter here when the caller requested more periods than the DMA can handle. We then calculate a buffer size from that larger block count and divide it by the the DMA maximum block count. So if max was 16 and the caller requested 17 and the period size isn't a multiple of 16, this won't divide well. Or is this meant for cases where the requested period count is a multiple of the block count, fitting multiple periods in a single DMA block?

Comment (abonislawski) [reply src/audio/dai-zephyr.c:501]: The only one real use case is limiting period_count to 1 for HDA DAI so more like as you said - fitting multiple period in a single DMA block, I thought we can make this more generic.
Division problem we can quickly solve with assert or ALIGN_DOWN but Im now not sure if this will result in a valid config, lets say 4*128 is still valid if requested 8*64?

**Review by tmleman — APPROVED** [review]

**Comment by marc-hb** [comment]
Re-running to clear https://sof-ci.01.org/sofpr/PR6823/build2944/boottest

https://sof-ci.01.org/sofpr/PR6823/build2944/devicetest/index.html is all green!

**Comment by marc-hb** [comment]
SOFCI TEST

File: src/audio/dai-zephyr.c (hunk 2 above)
Comment (abonislawski) [reply src/audio/dai-zephyr.c]: Updated with alignment check and period count decreasing, please take a look if it's not over complicated

File: src/audio/dai-zephyr.c (hunk 1 above)
Comment (kv2019i) [inline src/audio/dai-zephyr.c:452]: @abonislawski Can you update the commit message a bit as well and maybe add tag for the bug, i.e. "Link: https://github.com/thesofproject/sof/issues/6709" ?

**Review by kv2019i — APPROVED** [review]
Thanks! It would be good to update git commit message though.

File: src/audio/dai-zephyr.c (hunk 1 above)
Comment (abonislawski) [reply src/audio/dai-zephyr.c:452]: Updated
PR linked to 6709

**Comment by kv2019i** [comment]
Test results are good, including alsabat test now passing https://sof-ci.01.org/sofpr/PR6823/build3247/devicetest/index.html?model=TGLU_UP_HDA_IPC4ZPH&testcase=check-alsabat-headset-playback

There are multiple checkpatch warnings from long log/trace lines, but I think this is ok as it is not recommended to split string literals across multiple lines.

As this is a critical fix, proceeding with merge.
